{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5cdeec2a-5d2e-4ab6-9f57-b957c4d12383",
   "metadata": {},
   "source": [
    "# Build the phylogenetic tree\n",
    "***\n",
    "## Run the first steps of the panacotta pipeline :\n",
    "### I pangenome\n",
    "### II corepers\n",
    "### Prepare the files for the alignment\n",
    "### III align\n",
    "\n",
    "## Make phylogenetic tree with IQtree\n",
    "\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66a17da5-6ed2-4fa8-b39e-e5bb5cd4c716",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "\n",
    "def run_pangenome():\n",
    "    \"\"\"Run the pangenome command using singularity.\"\"\"\n",
    "    try:\n",
    "        cmd = [\n",
    "            \"singularity\", \"run\", \"/home/conchae/prediction_depolymerase_tropism/panacota.img\", \"pangenome\",\n",
    "            \"-l\", \"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/panacota_pangenome_list_v2.txt\",\n",
    "            \"-n\", \"Klebsiella_genomes_v2\",\n",
    "            \"-d\", \"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome_kp_kqp/protein_files\",\n",
    "            \"-o\", \"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome_kp_kqp\",\n",
    "            \"--threads\", \"0\", \"-v\"\n",
    "        ]\n",
    "\n",
    "        result = subprocess.run(cmd, check=True, capture_output=True, text=True)\n",
    "        print(f\"Pangenome output:\\n{result.stdout}\")\n",
    "\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error running pangenome: {e.stderr}\")\n",
    "    except Exception as ex:\n",
    "        print(f\"An unexpected error occurred: {ex}\")\n",
    "        \n",
    "\n",
    "def run_coregenome():\n",
    "    \"\"\"Run the coregenome command using singularity.\"\"\"\n",
    "    try:\n",
    "        cmd = [\n",
    "            \"singularity\", \"run\", \"/home/conchae/prediction_depolymerase_tropism/panacota.img\", \"corepers\",\n",
    "            \"-p\", \"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/PanGenome-Klebsiella_genomes.All.prt-clust-0.8-mode1-th80.lst\",\n",
    "            \"-t\", \"0.99\",\n",
    "            \"-o\", \"/home/conchae/prediction_depolymerase_tropism/panacota_core\"\n",
    "        ]\n",
    "\n",
    "        result = subprocess.run(cmd, check=True, capture_output=True, text=True)\n",
    "        print(f\"Coregenome output:\\n{result.stdout}\")\n",
    "\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error running coregenome: {e.stderr}\")\n",
    "    except Exception as ex:\n",
    "        print(f\"An unexpected error occurred: {ex}\")\n",
    "\n",
    "# Main execution\n",
    "if __name__ == \"__main__\":\n",
    "    run_pangenome()  # Run the pangenome command\n",
    "    run_coregenome()  # Run the coregenome command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85a7676d-5e2e-416f-8074-6780006a8bf8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "# Define paths\n",
    "path_klebsiella = \"/home/conchae/prediction_depolymerase_tropism\"\n",
    "path_prot_files = os.path.join(path_klebsiella, \"panacota_pangenome/protein_files\")\n",
    "path_data = os.path.join(path_klebsiella, \"panacota_align/data_prot_genes\")\n",
    "good_strains = open(os.path.join(path_klebsiella, \"panacota_pangenome/panacota_pangenome_list.txt\")).read().split(\"\\n\")\n",
    "\n",
    "\n",
    "# Utility function to format sequences in FASTA format\n",
    "def seq_fasta_format(seq):\n",
    "    \"\"\"Format the sequence for FASTA output.\"\"\"\n",
    "    if \"\\n\" in seq:\n",
    "        seq = \"\".join(seq.split(\"\\n\")[1:])\n",
    "    return \"\\n\".join([seq[i:i + 61] for i in range(0, len(seq), 61)])\n",
    "\n",
    "\n",
    "# Processing function to match proteins and genes\n",
    "def process_strain(strain, specie, faa_file, ffn_file):\n",
    "    \"\"\"Match protein sequences with gene sequences, and save the results in a .gen file.\"\"\"\n",
    "    gen_filepath = os.path.join(path_data, \"Genes\", f\"{strain}.gen\")\n",
    "    \n",
    "    # Check if the output file already exists\n",
    "    if not os.path.isfile(gen_filepath):\n",
    "        with open(gen_filepath, \"w\") as outfile_ffn:\n",
    "            used_ffn = []  # Track used gene sequences\n",
    "            for index_prot, seq_prot in enumerate(faa_file):\n",
    "                header = seq_prot.split(\"\\n\")[0]\n",
    "                fasta_faa = \"\".join(seq_prot.split(\"\\n\")[1:]).strip()\n",
    "                \n",
    "                # Search for matching gene sequences\n",
    "                for index_gen, seq_gen in enumerate(ffn_file):\n",
    "                    fasta_ffn = \"\".join(seq_gen.split(\"\\n\")[1:])\n",
    "                    translation = str(Seq(fasta_ffn).translate()).strip()\n",
    "\n",
    "                    # Check if the gene sequence matches the protein\n",
    "                    if len(fasta_ffn) % 3 == 0 and len(fasta_faa) == len(translation) and translation == fasta_faa:\n",
    "                        if ffn_file[index_gen].split(\"\\n\")[0] not in used_ffn:\n",
    "                            used_ffn.append(ffn_file[index_gen].split(\"\\n\")[0])\n",
    "                            outfile_ffn.write(f\">{header}\\n{seq_fasta_format(fasta_ffn)}\\n\")\n",
    "                            break  # Found the match, break out of the inner loop\n",
    "            print(f\"Finished processing strain: {strain}\")\n",
    "            \n",
    "\n",
    "# Function to process all species and strains\n",
    "def process_species_and_strains():\n",
    "    \"\"\"Iterate through species and strains, processing each strain.\"\"\"\n",
    "    for specie in os.listdir(path_klebsiella):\n",
    "        if specie.startswith(\"k\") and os.path.isdir(os.path.join(path_klebsiella, specie)):\n",
    "            strain_list = os.listdir(os.path.join(path_klebsiella, specie, \"refseq/bacteria\"))\n",
    "            for strain in random.sample(strain_list, len(strain_list)):\n",
    "                if strain in good_strains:\n",
    "                    faa_file_path = os.path.join(path_prot_files, f\"{strain}.prt\")\n",
    "                    ffn_file_path = os.path.join(path_klebsiella, specie, \"refseq/bacteria\", strain, \"prokka_annotation_all\", f\"{strain}.ffn\")\n",
    "\n",
    "                    # Check if both protein and gene files exist\n",
    "                    if os.path.isfile(faa_file_path) and os.path.isfile(ffn_file_path):\n",
    "                        faa_file = open(faa_file_path).read().split(\">\")[1:]\n",
    "                        ffn_file = open(ffn_file_path).read().split(\">\")[1:]\n",
    "\n",
    "                        process_strain(strain, specie, faa_file, ffn_file)\n",
    "                    else:\n",
    "                        print(f\"Missing files for strain: {strain}\")\n",
    "\n",
    "\n",
    "# Main execution\n",
    "if __name__ == \"__main__\":\n",
    "    process_species_and_strains()  # Process species and strains\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45aeea8a-b214-4451-87fa-8dbd1cd6fa89",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "\n",
    "def run_alignment():\n",
    "    \"\"\"Run the alignment command using singularity.\"\"\"\n",
    "    try:\n",
    "        cmd = [\n",
    "            \"singularity\", \"run\", \"/home/conchae/prediction_depolymerase_tropism/panacota.img\", \"align\",\n",
    "            \"-c\", \"/home/conchae/prediction_depolymerase_tropism/panacota_core/PersGenome_PanGenome-Klebsiella_genomes.All.prt-clust-0.8-mode1-th80.lst-all_0.99.lst\",\n",
    "            \"-l\", \"/home/conchae/prediction_depolymerase_tropism/panacota_pangenome/panacota_pangenome_list.txt\",\n",
    "            \"-n\", \"Klebsiella_genomes\",\n",
    "            \"-d\", \"/home/conchae/prediction_depolymerase_tropism/panacota_align/data_prot_genes\",\n",
    "            \"-o\", \"/home/conchae/prediction_depolymerase_tropism/panacota_align\",\n",
    "            \"--threads\", \"0\",\n",
    "            \"-v\", \"-P\"\n",
    "        ]\n",
    "\n",
    "        result = subprocess.run(cmd, check=True, capture_output=True, text=True)\n",
    "        print(f\"Alignment output:\\n{result.stdout}\")\n",
    "\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error running alignment: {e.stderr}\")\n",
    "    except Exception as ex:\n",
    "        print(f\"An unexpected error occurred: {ex}\")\n",
    "\n",
    "# Main execution\n",
    "if __name__ == \"__main__\":\n",
    "    run_alignment()  # Run the alignment command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "268ddc32-94e6-4c0f-ba90-517d8b888598",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_iqtree():\n",
    "    \"\"\"Run the IQ-TREE command to generate a phylogenetic tree.\"\"\"\n",
    "    try:\n",
    "        cmd = [\n",
    "            \"iqtree\",\n",
    "            \"-s\", \"/home/conchae/prediction_depolymerase_tropism/iqtree_local/Klebsiella_genomes.nucl.grp.aln\",\n",
    "            \"-m\", \"GTR+F+I\",\n",
    "            \"--prefix\", \"Klensiella_genomes_fixed\",\n",
    "            \"-B\", \"1000\",\n",
    "            \"-alrt\", \"1000\",\n",
    "            \"-t\", \"BIONJ\", \n",
    "            \"/home/conchae/prediction_depolymerase_tropism/iqtree_local/script_files/Klensiella_genomes_fixed.bionj\",\n",
    "            \"-nt\", \"AUTO\"\n",
    "        ]\n",
    "\n",
    "        result = subprocess.run(cmd, check=True, capture_output=True, text=True)\n",
    "        print(f\"IQ-TREE output:\\n{result.stdout}\")\n",
    "\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"Error running IQ-TREE: {e.stderr}\")\n",
    "    except Exception as ex:\n",
    "        print(f\"An unexpected error occurred: {ex}\")\n",
    "\n",
    "# Main execution\n",
    "if __name__ == \"__main__\":\n",
    "    run_iqtree()  # Run the IQ-TREE command"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:torch_geometric]",
   "language": "python",
   "name": "conda-env-torch_geometric-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
